packages/core/content-manager/server/src/controllers/relations.ts

```diff
@@ -167,133 +248,121 @@ export default {
+      // If it is a one to any relation, we query for the source content type
+      // This could be a component or a content type
+      modelUid = model;
+      filters = { id: currentEntityId };
+
+      if (sourceModelType === 'component') {
+        // If the source is a component, remove fields that are not present for components
+        fields = fields.filter(
+          (field: string) => !['documentId', 'locale', PUBLISHED_AT_ATTRIBUTE].includes(field)
+        );
       }
-    }
-
-    const targetedModel = strapi.getModel(attribute.target);
 
-    const modelConfig = isComponent
-      ? await getService('components').findConfiguration(modelSchema)
-      : await getService('content-types').findConfiguration(modelSchema);
+      populate = { [targetField]: { fields } };
+    }
 
-    const mainField = flow(
-      prop(`metadatas.${targetField}.edit.mainField`),
-      (mainField) => mainField || 'id',
-      (mainField) => sanitizeMainField(targetedModel, mainField, userAbility)
-    )(modelConfig);
+    const queryParams: Record<string, any> = {
+      fields,
+      filters,
+      sort,
+      populate,
+      page: ctx.request.query.page,
+      pageSize: ctx.request.query.pageSize,
+    };
 
-    const fieldsToSelect = uniq(['id', mainField, PUBLISHED_AT_ATTRIBUTE]);
+    const page = await strapi.entityService.findPage(modelUid, queryParams);
 
-    const queryParams = {
-      fields: fieldsToSelect,
-    };
+    // TODO simplify isOneToAny conditions
+    let results;
+    if (isOneToAny(attribute) && page.results.length === 1) {
+      let targetResult = page.results[0][targetField];
+      if (!targetResult) {
+        targetResult = [];
+      } else if (!Array.isArray(targetResult)) {
+        targetResult = [targetResult];
+      }
 
-    const permissionChecker = getService('permission-checker').create({
-      userAbility,
-      model: targetedModel.uid,
-    });
-    const permissionQuery = await permissionChecker.sanitizedQuery.read(queryParams);
-
-    if (isAnyToMany(attribute)) {
-      const res = await strapi.entityService.loadPages(
-        model,
-        { id },
-        targetField,
-        {
-          fields: ['id'],
-          ordering: 'desc',
-        } as any,
-        {
-          page: ctx.request.query.page,
-          pageSize: ctx.request.query.pageSize,
-        }
-      );
-      const ids = res.results.map((item: any) => item.id);
-
-      addFiltersClause(permissionQuery, { id: { $in: ids } });
-
-      const sanitizedRes = await strapi.entityService.loadPages(
-        model,
-        { id },
-        targetField,
-        {
-          ...permissionQuery,
-          ordering: 'desc',
-        } as any,
-        {
-          page: ctx.request.query.page,
-          pageSize: ctx.request.query.pageSize,
-        }
+      const getSortedResults = flow(
+        sortBy(
+          (value: any) => `${fieldsToSelect.map((field: any) => value?.[field] ?? null).join('-')}`
+        )
```

Comment:
why does it need to get sorted?

Reply:
Because if we follow this branch of the logic
```
      // If it is a one to any relation, we query for the source content type
      // This could be a component or a content type
      modelUid = model;
      filters = { id: currentEntityId };

      if (sourceModelType === 'component') {
        // If the source is a component, remove fields that are not present for components
        fields = fields.filter(
          (field: string) => !['documentId', 'locale', PUBLISHED_AT_ATTRIBUTE].includes(field)
        );
      }

      populate = { [targetField]: { fields } };
```

We are not sorting the response from the entityService.

Which is because I query by the source type and I wasn't sure how to sort the values in a relational property?